File: fingerprintmatcher/src/main/java/com/simprints/fingerprintmatcher/algorithms/simafis/JNILibAfis.java
package com.simprints.fingerprintmatcher.algorithms.simafis;

import com.simprints.fingerprintmatcher.algorithms.simafis.models.SimAfisPerson;

import java.nio.ByteBuffer;
import java.util.List;

public class JNILibAfis {
    static
    {
        System.loadLibrary( "simmatcherwrapper" );
    }

    /*
     * We use a class initializer to allow the native code to cache some
     * field offsets. This native function looks up and caches interesting
     * class/field/method IDs. Throws on failure.
     */
    private static native boolean nativeInit();

    static
    {
        nativeInit();
    }

    public static native int getNbCores();

    public static native float verify(ByteBuffer probe, ByteBuffer candidate);

    public static native float[] identify(SimAfisPerson probe, List<SimAfisPerson> candidates, int nbThreads);

}

File: fingerprint/infra/simafis-wrapper/src/main/java/com/simprints/fingerprint/infra/simafiswrapper/JNILibAfis.kt
package com.simprints.fingerprint.infra.simafiswrapper

import com.simprints.fingerprint.infra.simafiswrapper.models.SimAfisPerson
import java.nio.ByteBuffer

object JNILibAfis : JNILibAfisInterface {
    init {
        System.loadLibrary("simmatcherwrapper")
        nativeInit()
    }

    external override fun nativeInit(): Boolean

    external override fun getNbCores(): Int

    external override fun verify(
        probe: ByteBuffer,
        candidate: ByteBuffer,
    ): Float

    external override fun identify(
        probe: SimAfisPerson,
        candidates: List<SimAfisPerson>,
        nbThreads: Int,
    ): FloatArray
}
